pandas/core/reshape/util.py

```diff
@@ -39,6 +39,10 @@ def cartesian_product(X):
     lenX = np.fromiter((len(x) for x in X), dtype=np.intp)
     cumprodX = np.cumproduct(lenX)
 
+    msg = "Product space too large to allocate arrays!"
```

Comment:
Nit: could you put this inside the condition since it only matters in that case (or simply inside the ValueError)?